smarts/core/sumo_road_network.py

```diff
@@ -536,17 +560,11 @@ def world_coord_from_offset(self, lane: Lane, offset) -> np.ndarray:
 
     def point_is_within_road(self, point):
         # XXX: Not robust around junctions (since their shape is quite crude?)
-        # # We use a small radius 10 factor in some forgiveness for crude SUMO distance checks
-        # for lane, dist in self.nearest_lanes(point[:2], radius=10):
-        #     shape = SumoRoadNetwork.buffered_lane_or_edge(lane, lane.getWidth() * 1)
-        #     if sumolib.geomhelper.isWithin(point, shape):
-        #         return True
-        #
-        # return False
-
         radius = max(5, 2 * self._default_lane_width)
-        nearest_lanes = self.nearest_lanes(point[:2], radius=radius)
-        return any(dist < 0.5 * nl.getWidth() + 1e-1 for nl, dist in nearest_lanes)
+        for nl, dist in self.nearest_lanes(point[:2], radius=radius):
+            if dist < 0.5 * nl.getWidth() + 1e-1:
+                return True
+        return False
```

Comment:
Actually, this was also part of that attempted optimization that I added and then rolled back (described in my comment on `_vehicle_is_on_shoulder()`).

However, `any` is inherently less efficient when you only care about one instance, even in list comprehensions, so I kept this change in.  For example, consider the following:
```python
def foo(x):
    return x % 3 == 0
any([foo(x) for x in range(100)])
```
Here, `foo()` is called 100 times, even though it only needs to be called 3 times to decide the truth value of the expression.

Similarly, in the original code, the expression `dist < 0.5 * nl.getWidth() + 1e-1` can potentially be called more times than necessary.  This method is called at least 5 times per agent per step which, for traffic history replay at least, adds up.  (Although the real thing I was originally trying to optimize/avoid was so many calls to `getNeighboringLanes()`, I just did this along the way and then didn't roll it back.)  It did make a noticeable (albeit small percentage) difference in my profiling.

Reply:
I think that would be true if it was a list being supplied to `any` but it was a generator so it should still short circuit.

```
def foo(x): 
   return x % 3 == 2 
any(foo(x) for x in range(100))
```
which is semantically the same as:
```
def foo(): 
   for x in range(100):   
      yield x % 3 == 2 
any(foo())
```

Reply:
Yes, you're right about the generator vs. the list comprehension, sorry.  So based on that, I would _expect_ there would be no difference between the two and, in that case, I would prefer the shorter `any` version.  _HOWEVER..._

I'm still confused about this because I really did see a performance bump after the change.  So I just went back and did the test again and, surprisingly enough, it's still there!

Here's a table of results of `single_agent.py` run on the `intersections/4lane` scenario with the `any` version:
```
╭────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────╮
│            Episode │      Sim T / Sys T │     Sim T / Wall T │        Total Steps │     Steps / SysSec │        Steps / Sec │       Scenario Map │    Scenario Routes │     Mission (Hash) │             Scores │
├────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┤
│               0/10 │               3.28 │               7.47 │                156 │              32.82 │              74.74 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               1/10 │               3.62 │               8.33 │                156 │              36.20 │              83.29 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               2/10 │               3.84 │               8.80 │                156 │              38.36 │              87.99 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               3/10 │               3.84 │               8.73 │                156 │              38.41 │              87.33 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               4/10 │               3.82 │               8.72 │                156 │              38.20 │              87.22 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               5/10 │               3.61 │               8.23 │                156 │              36.14 │              82.27 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               6/10 │               3.74 │               8.51 │                156 │              37.36 │              85.13 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               7/10 │               3.82 │               8.80 │                156 │              38.16 │              87.98 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               8/10 │               3.57 │               8.25 │                156 │              35.66 │              82.46 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
│               9/10 │               3.52 │               8.14 │                156 │              35.17 │              81.42 │              4lane │      basic.rou.xml │ -18965767758539274 │ 192.34 - Agent-007 │
╰────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────╯
Elapsed Time: 44.69 seconds.  Collected 4758 stack frames (2256 unique)
```
And here's the same thing with the nested loops:
```
╭────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────┬────────────────────╮
│            Episode │      Sim T / Sys T │     Sim T / Wall T │        Total Steps │     Steps / SysSec │        Steps / Sec │       Scenario Map │    Scenario Routes │     Mission (Hash) │             Scores │
├────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┼────────────────────┤
│               0/10 │               3.39 │               7.60 │                156 │              33.91 │              76.02 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               1/10 │               3.83 │               8.77 │                156 │              38.32 │              87.72 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               2/10 │               3.28 │               7.92 │                156 │              32.80 │              79.15 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               3/10 │               3.88 │               8.84 │                156 │              38.76 │              88.36 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               4/10 │               3.79 │               8.86 │                156 │              37.90 │              88.58 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               5/10 │               3.65 │               8.36 │                156 │              36.51 │              83.64 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               6/10 │               3.83 │               8.81 │                156 │              38.29 │              88.10 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               7/10 │               3.89 │               8.93 │                156 │              38.85 │              89.34 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               8/10 │               3.87 │               8.88 │                156 │              38.67 │              88.79 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
│               9/10 │               3.74 │               8.53 │                156 │              37.44 │              85.27 │              4lane │      basic.rou.xml │ -44837943675001737 │ 192.34 - Agent-007 │
╰────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────┴────────────────────╯
Elapsed Time: 44.14 seconds.  Collected 4656 stack frames (2247 unique)
```

Note that to create that table, I modified `util/episodes.py` to add two columns based on the elapsed CPU time (obtained using `time.process_time`).  The main column to pay attention to then is `Sim T / Sys T`.  In the first table, it's average is  `3.666` and in the second it's `3.715`, which is a `1.3%` speedup.  I saw similar speedups on the `loop` scenario as well.

I'm going to dig a little more to try to understand what's going on.  Will report back...

Reply:
Just to clarify, the _only_ difference in the code between the two tables above is:
```python
        nearest_lanes = self.nearest_lanes(point[:2], radius=radius)
        return any(dist < 0.5 * nl.getWidth() + 1e-1 for nl, dist in nearest_lanes)
```
vs.
```python
        for nl, dist in self.nearest_lanes(point[:2], radius=radius):
            if dist < 0.5 * nl.getWidth() + 1e-1:
                return True
        return False
```

Reply:
Upon some thought and experimentation, I think there must be some overhead to set up a generator in the `any` case.

Reply:
That is interesting, I can only assume there is bunch of hidden complexity behind `any` even though it should be functionally the same. I will keep that in mind for the future.